src/tss2-tcti/tcti-libtpms.c

```diff
@@ -0,0 +1,969 @@
+{
+    pid_t thread_id = syscall(__NR_gettid);
+    int index;
+
+    /* return if thread is already registered */
+    for (size_t i = 0; i < ARRAY_LEN(current_libtpms_tctis); i++) {
+        if (current_libtpms_tctis[i].thread_id == thread_id) {
+            return i;
+        }
+    }
+
+    /* did not find thread in the list, for write access we need to lock */
+    pthread_mutex_lock(&current_libtpms_tctis_lock);
+
+    /* add new entry to list */
+    for (size_t i = 0; i < ARRAY_LEN(current_libtpms_tctis); i++) {
+        if (current_libtpms_tctis[i].thread_id == 0) {
+            current_libtpms_tctis[i].thread_id = thread_id;
+            current_libtpms_tctis[i].tcti_libtpms = tcti_libtpms;
+
+            LOG_TRACE("Registered thread %d in current_libtpms_tctis[%zu]",
+                      thread_id, i);
+            index = i;
+            goto cleanup_mutex;
+        }
+    }
+
+    LOG_ERROR("Failed to register thread %d because there is not more space in list.",
+              thread_id);
+
+    index = -1;
+
+cleanup_mutex:
+    pthread_mutex_unlock(&current_libtpms_tctis_lock);
+
+    return index;
+}
+
+/*
+ * Set the currently active libtpms TCTI for this thread to tcti_libtpms.
+ */
+static TSS2_RC
+tcti_libtpms_set_current_tcti(TSS2_TCTI_LIBTPMS_CONTEXT *tcti_libtpms)
+{
+    int index = tcti_libtpms_register_get_thread_index(tcti_libtpms);
+    if (index < 0) {
+        return TSS2_TCTI_RC_GENERAL_FAILURE;
+    }
+
+    /* set its current tcti to this */
+    current_libtpms_tctis[index].tcti_libtpms = tcti_libtpms;
+
+    return TPM2_RC_SUCCESS;
+}
+
+/*
+ * Reset the currently active libtpms TCTI for this thread.
+ */
+static TSS2_RC
+tcti_libtpms_clear_current_tcti(void)
+{
+    return tcti_libtpms_set_current_tcti(NULL);
+}
+
+/*
+ * Get the currently active libtpms TCTI for this thread.
+ */
+static TSS2_RC
+tcti_libtpms_get_current_tcti(TSS2_TCTI_LIBTPMS_CONTEXT **tcti_libtpms)
+{
+    int index = tcti_libtpms_register_get_thread_index(NULL);
+    if (index < 0) {
+        return TSS2_TCTI_RC_GENERAL_FAILURE;
+    }
+
+    *tcti_libtpms = current_libtpms_tctis[index].tcti_libtpms;
+    if (tcti_libtpms == NULL) {
+        LOG_ERROR("No TCTI registered for this thread: %ld", syscall(__NR_gettid));
+        return TSS2_TCTI_RC_GENERAL_FAILURE;
+    }
+
+    return TPM2_RC_SUCCESS;
+}
+
+/*
+ * Map the state file for this context into memory and allocate disk space. The
+ * file descriptor is closed again. Once this context reaches the end of its
+ * lifetime, the memory must be unmapped and the file must be truncated to its
+ * real size (rather than the allocated size).
+ */
+static TSS2_RC
+tcti_libtpms_map_state_file(TSS2_TCTI_LIBTPMS_CONTEXT *tcti_libtpms)
+{
+    TSS2_RC rc;
+    int ret;
+    int state_fd = -1;
+    ssize_t file_len = 0;
+
+    tcti_libtpms->state_mmap_len = STATE_MMAP_CHUNK_LEN;
+
+    if (tcti_libtpms->state_path != NULL) {
```

Comment:
Looks like this function should be a noop() if tcti_libtpms->state_path == NULL so let's check this condition first here and return a SUCCESS if it is. Same for tcti_libtpms_load_state().
This will allow to simplify the code in Tss2_Tcti_Libtpms_Init() where these functions are called.

Reply:
Yes, that's also a way to do it. Will do.